matcher.go

```diff
@@ -1,43 +1,61 @@
 package gostatsd
 
 import (
+	"regexp"
 	"strings"
 )
 
 type StringMatch struct {
 	test        string
 	invertMatch bool
 	prefixMatch bool
+	isRegex     bool
+	regex       *regexp.Regexp
 }
 
 type StringMatchList []StringMatch
 
 func NewStringMatch(s string) StringMatch {
+
 	invert := false
 	if strings.HasPrefix(s, "!") {
 		invert = true
 		s = s[1:]
 	}
 
-	prefix := false
-	if strings.HasSuffix(s, "*") {
-		prefix = true
-		s = s[0 : len(s)-1]
+	regex := false
+	var compiledRegex *regexp.Regexp = nil
+	if strings.HasPrefix(s, "regex:") {
+		regex = true
+		s = s[6:]
+		compiledRegex, _ = regexp.Compile(s)
 	}
-
+	prefix := false
+        if !regex {
+		if strings.HasSuffix(s, "*") {
+			prefix = true
+			s = s[0 : len(s)-1]
+		}
+        }
 	return StringMatch{
 		test:        s,
 		invertMatch: invert,
 		prefixMatch: prefix,
+		isRegex:     regex,
+		regex:       compiledRegex,
 	}
 }
 
 // Match indicates if the provided string matches the criteria for this StringMatch
 func (sm StringMatch) Match(s string) bool {
-	if sm.prefixMatch {
-		return strings.HasPrefix(s, sm.test) != sm.invertMatch
+	if sm.isRegex {
+		return (sm.regex.MatchString(s)) != sm.invertMatch
+	} else {
```

Comment:
`else` may be omitted (Go style) because of the return in the `if` block.

Reply:
In this particular case, I'd personally keep the `else` in, because I think the code flows better. Although since you've got me thinking about it now, a `switch` is probably even clearer because it's more explicit and regular in the flow:
```
switch {
case sm.isRegex:
  return sm.regex.MatchString(s) != sm.invertMatch
case sm.prefixMatch:
  return strings.HasPrefix(s, sm.test) != sm.invertMatch
default: // exact match
  return (s == sm.test) != sm.invertMatch
}
```